gobblin-core/src/main/java/org/apache/gobblin/converter/grok/GrokToJsonConverter.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gobblin.converter.grok;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStreamReader;
+import java.io.UnsupportedEncodingException;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonNull;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import io.thekraken.grok.api.Grok;
+import io.thekraken.grok.api.Match;
+import io.thekraken.grok.api.exception.GrokException;
+
+import org.apache.gobblin.configuration.WorkUnitState;
+import org.apache.gobblin.converter.Converter;
+import org.apache.gobblin.converter.DataConversionException;
+import org.apache.gobblin.converter.SchemaConversionException;
+import org.apache.gobblin.converter.SingleRecordIterable;
+
+
+/**
+ * GrokToJsonConverter accepts already deserialized text row, String, where you can use.
+ *
+ * Converts Text to JSON based on Grok pattern. Schema is represented by the form of JsonArray same interface being used by CsvToJonConverter.
+ * Each text record is represented by a String.
+ * The converter only supports Grok patterns where every group is named because it uses the group names as column names.
+ *
+ * The following config properties can be set:
+ * The grok pattern to use for the conversion:
+ * converter.grok_to_json.pattern ="^%{IPORHOST:clientip} (?:-|%{USER:ident}) (?:-|%{USER:auth}) \[%{HTTPDATE:timestamp}\] \"(?:%{WORD:verb} %{NOTSPACE:request}(?: HTTP/%{NUMBER:httpversion})?|-)\" %{NUMBER:response} (?:-|%{NUMBER:bytes})"
+ *
+ * Path to the file which contains the base grok patterns which can be used in the converter's GROK pattern (if not set it will use the default ones):
+ * converter.grok_to_json.base_patterns=
+ *
+ * Treat empty string as json null value. If it is true the empty string will be converted to json null:
+ * converter.grok_to_json.empty_as_null=true
```

Comment:
Can we merge the configuration `converter.grok_to_json.empty_as_null` and `converter.grok_to_json.null_string`? We can have something like
`converter.grokToJsonConverter.stringsAsNull=,None`?

Also, the configuration key names in the comments should be updated.

Reply:
Sorry, I missed those but I fixed them now.
I also extended empty as null to treat string which only contains whitespaces as empty.
I still think there is value in having null string and empty as null separate as there can be usecase where people have a specific null string like '-' or 'null' but they would like to treat strings which only contains whitespaces or no character as null. 
The idea is pretty similar like here: http://docs.aws.amazon.com/redshift/latest/dg/copy-parameters-data-conversion.html#copy-emptyasnull and in our environment are using this option and the null string one as well separately.
If you guys still don't see the value, I'm happy to remove it but I think it is useful and even it could be useful at the csv converter.
What do you think?

Reply:
I do see the value in making certain strings as null, but I don't see the value in using 2 separate configuration keys to achieve that. Only one configuration would suffice. Another idea is to have a configuration like `converter.grokToJsonConverter.nullStringRegexes`, which is a list of regex patterns identifying strings that should be replaced as null. Check out `DatasetFilterUtils` to see how a list of patterns is used.